File: src/utils/typeFilter.ts
import type { Position } from '../types'

export type AssetType = Position['type']

/** Display order for chips and labels. Mirrors `TYPE_ORDER` in core/calc/TypeFilter.kt. */
export const TYPE_ORDER: AssetType[] = ['stock', 'etf', 'fund', 'commodity', 'crypto']

const LABELS: Record<AssetType, string> = { stock: 'Stock', etf: 'ETF', fund: 'Fund', commodity: 'Commodity', crypto: 'Crypto' }
export const typeLabel = (t: AssetType): string => LABELS[t]

/**
 * The filter actually in force: the stored selection minus types no longer held, so a stale
 * stored filter can never hide the whole chart. Empty means All.
 */
export function effectiveTypeFilter(selected: AssetType[], held: AssetType[]): Set<AssetType> {
  const heldSet = new Set(held)
  return new Set(TYPE_ORDER.filter((t) => selected.includes(t) && heldSet.has(t)))
}

export function filterPositionsByType(positions: Position[], filter: Set<AssetType>): Position[] {
  return filter.size === 0 ? positions : positions.filter((p) => filter.has(p.type))
}

/** Adds or removes one type; the result is in display order, and empty means All. */
export function toggleType(selected: AssetType[], t: AssetType): AssetType[] {
  const next = selected.includes(t) ? selected.filter((x) => x !== t) : [...selected, t]
  return TYPE_ORDER.filter((x) => next.includes(x))
}

export function typeFilterLabel(filter: Set<AssetType>): string {
  return TYPE_ORDER.filter((t) => filter.has(t)).map(typeLabel).join(' + ')
}

export function parseStoredTypes(raw: string | null): AssetType[] {
  if (!raw) return []
  try {
    const v: unknown = JSON.parse(raw)
    return Array.isArray(v) ? TYPE_ORDER.filter((t) => v.includes(t)) : []
  } catch {
    return []
  }
}

File: android/core/calc/src/main/kotlin/com/stocktracker/core/calc/TypeFilter.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.Position
import com.stocktracker.core.model.PositionType

/** Display order for chips and labels. Mirrors `TYPE_ORDER` in src/utils/typeFilter.ts. */
val TYPE_ORDER: List<PositionType> = listOf(
    PositionType.STOCK, PositionType.ETF, PositionType.FUND, PositionType.COMMODITY, PositionType.CRYPTO,
)

fun typeLabel(type: PositionType): String = when (type) {
    PositionType.STOCK -> "Stock"
    PositionType.ETF -> "ETF"
    PositionType.FUND -> "Fund"
    PositionType.COMMODITY -> "Commodity"
    PositionType.CRYPTO -> "Crypto"
}

/**
 * The filter actually in force: the stored selection minus types no longer held, so a stale
 * stored filter can never hide the whole chart. Empty means All.
 */
fun effectiveTypeFilter(selected: Set<PositionType>, held: Set<PositionType>): Set<PositionType> =
    TYPE_ORDER.filter { it in selected && it in held }.toCollection(LinkedHashSet())

fun filterPositionsByType(positions: List<Position>, filter: Set<PositionType>): List<Position> =
    if (filter.isEmpty()) positions else positions.filter { it.type in filter }

/** Adds or removes one type; the result is in display order, and empty means All. */
fun toggleType(selected: Set<PositionType>, type: PositionType): Set<PositionType> {
    val next = if (type in selected) selected - type else selected + type
    return TYPE_ORDER.filter { it in next }.toCollection(LinkedHashSet())
}

fun typeFilterLabel(filter: Set<PositionType>): String =
    TYPE_ORDER.filter { it in filter }.joinToString(" + ") { typeLabel(it) }
